internal/pkg/istio/config_factory_test.go

```diff
@@ -0,0 +1,306 @@
+			},
+			Spec: istionetv1alpha3.Gateway{
+				Selector: map[string]string{"app": "federation-ingress-gateway"},
+				Servers: []*istionetv1alpha3.Server{{
+					Hosts: []string{
+						"a.ns2.svc.cluster.local",
+						"b.ns1.svc.cluster.local",
+						"federation-discovery-service-east.istio-system.svc.cluster.local",
+					},
+					Port: &istionetv1alpha3.Port{
+						Number:   443,
+						Name:     "tls",
+						Protocol: "TLS",
+					},
+					Tls: &istionetv1alpha3.ServerTLSSettings{
+						Mode: istionetv1alpha3.ServerTLSSettings_AUTO_PASSTHROUGH,
+					},
+				}},
+			},
+		},
+	}, {
+		name:             "federation-ingress-gateway should always expose FDS",
+		existingServices: []*corev1.Service{},
+		expectedGateway: &v1alpha3.Gateway{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      "federation-ingress-gateway",
+				Namespace: "istio-system",
+				Labels:    map[string]string{"federation.istio-ecosystem.io/peer": "todo"},
+			},
+			Spec: istionetv1alpha3.Gateway{
+				Selector: map[string]string{"app": "federation-ingress-gateway"},
+				Servers: []*istionetv1alpha3.Server{{
+					Hosts: []string{
+						"federation-discovery-service-east.istio-system.svc.cluster.local",
+					},
+					Port: &istionetv1alpha3.Port{
+						Number:   443,
+						Name:     "tls",
+						Protocol: "TLS",
+					},
+					Tls: &istionetv1alpha3.ServerTLSSettings{
+						Mode: istionetv1alpha3.ServerTLSSettings_AUTO_PASSTHROUGH,
+					},
+				}},
+			},
+		},
+	}}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			client := fake.NewSimpleClientset()
+			informerFactory := informers.NewSharedInformerFactory(client, 0)
+			serviceInformer := informerFactory.Core().V1().Services().Informer()
+			serviceLister := informerFactory.Core().V1().Services().Lister()
+			stopCh := make(chan struct{})
+			informerFactory.Start(stopCh)
+
+			for _, svc := range tc.existingServices {
+				if _, err := client.CoreV1().Services(svc.Namespace).Create(context.Background(), svc, v1.CreateOptions{}); err != nil {
+					t.Fatalf("failed to create service %s/%s: %v", svc.Name, svc.Namespace, err)
+				}
+			}
+
+			serviceController, err := informer.NewResourceController(serviceInformer, corev1.Service{})
+			if err != nil {
+				t.Fatalf("error creating serviceController: %v", err)
+			}
+			serviceController.RunAndWait(stopCh)
+
+			factory := NewConfigFactory(exportConfig, serviceLister, fds.NewImportedServiceStore(), "istio-system")
+			actual, err := factory.IngressGateway()
+			if err != nil {
+				t.Errorf("got unexpected error: %s", err)
+			}
+			if !reflect.DeepEqual(actual, tc.expectedGateway) {
+				t.Errorf("got unexpected result:\nexpected:\n%v\ngot:\n%v", tc.expectedGateway, actual)
+			}
+		})
+	}
+}
+
+func TestEnvoyFilters(t *testing.T) {
+	testCases := []struct {
+		name                     string
+		localIngressType         config.IngressType
+		existingServices         []*corev1.Service
+		expectedEnvoyFilterFiles []string
```

Comment:
In case of EnvoyFilters it's better to compare results to the serialized objects, because creating proper patch values require to use tricky protobuf serialization, and we would have to copy function from the config factory, and such a test would not make any value to me.

Reply:
I was actually thinking about bringing some assertion library and use `jq` for more readable tests. But I don't have a good example yet to show value in improved readability so I will ponder more about it.